vllm/lora/punica.py

```diff
@@ -1,206 +1,260 @@
+from vllm.lora.ops.sgmv_expand import sgmv_expand
+from vllm.lora.ops.sgmv_expand_slice import sgmv_expand_slice
+from vllm.lora.ops.sgmv_shrink import sgmv_shrink
 
+_PARAMS_CACHE: Dict[int, Tuple] = {}
 
-def _check_punica_support():
-    if ops.is_custom_op_supported("_punica_C::dispatch_bgmv"):
-        return
 
-    if torch.cuda.get_device_capability() < (8, 0):
-        raise ImportError(
-            "punica LoRA kernels require compute capability >= 8.0")
-    else:
-        raise ImportError(
-            "punica LoRA kernels could not be imported. If you built vLLM "
-            "from source, make sure VLLM_INSTALL_PUNICA_KERNELS=1 env var "
-            "was set.")
+def _compute_params(
+    token_lora_tensor: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, int, int, ]:
+    """
+    Get the information required for the sgmv kernel.
+    """
+    pointer = token_lora_tensor.data_ptr()
+    if pointer not in _PARAMS_CACHE:
+        lora_indices_tensor, seq_length_tensor = torch.unique_consecutive(
+            token_lora_tensor, return_counts=True)
+        cum_result = torch.cumsum(seq_length_tensor, dim=0)
+        b_seq_start_tensor = torch.zeros_like(seq_length_tensor)
+        b_seq_start_tensor[1:].copy_(cum_result[:-1])
+        max_length = seq_length_tensor.max().item()
+        batch_size = lora_indices_tensor.size(0)
+        _PARAMS_CACHE[pointer] = (
+            b_seq_start_tensor,
+            seq_length_tensor,
+            lora_indices_tensor,
+            batch_size,
+            max_length,
+        )
+    return _PARAMS_CACHE[pointer]
 
 
-def bgmv(
+def reset_params_cache():
+    """At the beginning of the prefilling stage, we need  clear the
+    cache explicitly
+    """
+    #TODO release gpu memory
+    torch.cuda.empty_cache()
+    _PARAMS_CACHE.clear()
+
+
+def _get_prefilling_params(token_lora_tensor: torch.Tensor,
+                           cache_clear: bool = False):
+    if cache_clear:
+        reset_params_cache()
+    return _compute_params(token_lora_tensor)
+
+
+def add_shrink(
     y: torch.Tensor,
     x: torch.Tensor,
     w_t_all: torch.Tensor,
-    indicies: torch.LongTensor,
+    lora_indices_tensor: torch.Tensor,
     layer_idx: int,
     scale: float,
+    is_prefilling: bool,
+    cache_clear: bool = False,
 ):
     """
-    Semantics:
-      y[i] += (
-          x[i].unsqueeze(0)
-          @ w_t_all[indices[i], layer_idx, :, :].transpose(-1, -2)
-          * scale
-        ).squeeze(0)
-
-    Args:
-      y: Shape: `[B, H2]`. Output vectors. Will be changed in-place.
-      x: Shape: `[B, H1]`. Input vectors.
-      w_t_all: Shape: `[None, L, H2, H1]`. All of the transposed weight
-        matrices.
-      indicies: Shape: `[B]`. Indices of the weight matrices.
-      layer_idx: Layer index of the weight matrices.
-      scale: Scaling factor.
+    y=x@w_t_all
+    When `is_prefilling` is True, will launch `sgmv_shrink`
```

Comment:
I think it makes more sense to break those into shrink_decode & shrink_prefill instead of using `is_prefilling` arg (same for expand)

Reply:
Thanks, we should modify the code like the following pseudocode, is that correct?
```python
def shrink_prefill:
    pass
def shrink_decode:
    pass
def add_shrink:
    if is_prefilling:
        shrink_prefill(...)
    else:
        shrink_decode(...)
```

Reply:
yes, I think so